launcher/modplatform/modrinth/ModrinthPackExportTask.cpp

```diff
@@ -0,0 +1,323 @@
+{
+    setStatus(tr("Searching for files..."));
+    setProgress(0, 0);
+    collectFiles();
+}
+
+bool ModrinthPackExportTask::abort()
+{
+    if (task != nullptr) {
+        task->abort();
+        task = nullptr;
+        emitAborted();
+        return true;
+    }
+
+    if (buildZipFuture.isRunning()) {
+        buildZipFuture.cancel();
+        // NOTE: Here we don't do `emitAborted()` because it will be done when `buildZipFuture` actually cancels, which may not occur immediately.
+        return true;
+    }
+
+    return false;
+}
+
+void ModrinthPackExportTask::collectFiles()
+{
+    setAbortable(false);
+    QCoreApplication::processEvents();
+
+    files.clear();
+    if (!MMCZip::collectFileListRecursively(instance->gameRoot(), nullptr, &files, filter)) {
+        emitFailed(tr("Could not search for files"));
+        return;
+    }
+
+    pendingHashes.clear();
+    resolvedFiles.clear();
+
+    if (mcInstance) {
+        mcInstance->loaderModList()->update();
+        connect(mcInstance->loaderModList().get(), &ModFolderModel::updateFinished, this, &ModrinthPackExportTask::collectHashes);
+    } else
+        collectHashes();
+}
+
+void ModrinthPackExportTask::collectHashes()
+{
+    for (const QFileInfo& file : files) {
+        QCoreApplication::processEvents();
+
+        const QString relative = gameRoot.relativeFilePath(file.absoluteFilePath());
+        // require sensible file types
+        if (!std::any_of(PREFIXES.begin(), PREFIXES.end(),
+                         [&relative](const QString& prefix) { return relative.startsWith(prefix + QDir::separator()); }))
+            continue;
+        if (!std::any_of(FILE_EXTENSIONS.begin(), FILE_EXTENSIONS.end(), [&relative](const QString& extension) {
+                return relative.endsWith('.' + extension) || relative.endsWith('.' + extension + ".disabled");
+            })) {
+            continue;
+        }
+
+        QCryptographicHash sha512(QCryptographicHash::Algorithm::Sha512);
+
+        QFile openFile(file.absoluteFilePath());
+        if (!openFile.open(QFile::ReadOnly)) {
+            qWarning() << "Could not open" << file << "for hashing";
+            continue;
+        }
+
+        const QByteArray data = openFile.readAll();
+        if (openFile.error() != QFileDevice::NoError) {
+            qWarning() << "Could not read" << file;
+            continue;
+        }
+        sha512.addData(data);
+
+        auto allMods = mcInstance->loaderModList()->allMods();
+        if (auto modIter = std::find_if(allMods.begin(), allMods.end(), [&file](Mod* mod) { return mod->fileinfo() == file; });
+            modIter != allMods.end()) {
+            const Mod* mod = *modIter;
+            if (mod->metadata() != nullptr) {
+                QUrl& url = mod->metadata()->url;
+                // most likely some of these may be from curseforge
+                if (!url.isEmpty() && BuildConfig.MODRINTH_MRPACK_HOSTS.contains(url.host())) {
```

Comment:
can't you just check `metadata()->provider` to see if it's from Modrinth?

Reply:
I somehow managed to screw this up :P sorry
that wouldn't account for GitHub and GitLab files

Reply:
Doesn't Modrinth accept those too? Or is this if statement there only to filter out the CurseForge mods, since those can't be directly download per Modrinth's (well, CF's really) policy? If that's the case, I'd rather have it check for something among the lines of `metadata()->provider != ModPlatform::ResourceProvider::FLAME`.

Reply:
Well the goal is to keep *GitHub and GitLab* files which won't be marked as Modrinth files

Reply:
I see. I think it's quite weird for us, as a third-party in relation to Modrinth, to uphold their whitelist on a modpack format that they, at times overly explicitly, state need not be tied to their service.

I understand the logic though. I would just like to ask you to document a bit better that in the code. The way it's written, it's not easy to understand that the check is there so that mrpacks exported from Prism are compatible with Modrinth's upload policies.